### DISCUSSION TITLE
Infinitely Running Callbacks and Resetting Retry Attempts

### DISCUSSION THREAD

**Discussion by ghost** [App-vNext/Polly#1969] [Q&A]
Hello Polly Community,

currently I'm using Polly's resilience pipeline with a callback method that runs *infinitely*. My goal is to recover from network failures that may occur inside an event loop, i.e. I want to try to restablish the connections and then restart the event loop.

```
new ResiliencePipelineBuilder()
    .AddRetry(new RetryStrategyOptions()
    {
        BackoffType = DelayBackoffType.Exponential,    
        MaxDelay = TimeSpan.FromMinutes(10),           
        Delay = TimeSpan.FromSeconds(8),               
        MaxRetryAttempts = int.MaxValue,               
        UseJitter = true
    })
    .Build();
```

However, I've encountered a specific challenge related to the retry strategy, especially with exponential backoff. Once the retry interval reaches its maximum (for example, a 10-minute delay after several consecutive failures), it does not reset even if the system recovers and remains healthy for an extended period. This behavior leads to less-than-ideal retry intervals when new failures occur after the system has been stable.

I'm wondering if Polly's resilience pipelines are intended for use with infinitely running processes?

Is there a built-in mechanism to reset the retry strategy after the system remains healthy for a sufficient duration? Specifically, I'm looking for a way to reset the exponential backoff to its initial state after the system has demonstrated stability for a certain period, ensuring that the retry intervals are appropriate for the current state of the system.

I would be very grateful for any insights, recommendations or examples of how Polly can be used effectively in such contexts.

**Comment by peter-csala** [comment]
Are you using circuit breaker as well?

**Reply by ghost** [reply]
No, if I'm not mistaken, the circuit breaker is intended for use with *consecutive* calls to `ExecuteAsync`. My specific use case involves calling `ExecuteAsync` once and leveraging the resilience pipeline to ensure that the process (the callback method) runs indefinitely.

**Reply by peter-csala** [reply]
No. It was true in case of V7, but that's not the case anymore for V8. The V8's Circuit Breaker works similarly like V7's Advanced Circuit Breaker. This is called out in the migration doc: https://www.pollydocs.org/migration-v8.html#circuit-breaker-in-v8.

V8's circuit breaker opens if we had at least `MinimumThroughput` number of requests during the `SamplingDuration` and the success/failure ratio is greater or equal than `FailureRatio`.
